src/mixins/java/org/spongepowered/common/mixin/core/server/network/ServerGamePacketListenerImplMixin.java

```diff
@@ -362,10 +356,112 @@
                 packetInAccessor.accessor$x(toPosition.x());
                 packetInAccessor.accessor$y(toPosition.y());
                 packetInAccessor.accessor$z(toPosition.z());
+                if (packetInAccessor.accessor$hasRot()) {
+                    // this is intended, see the from rotation line above.
+                    packetInAccessor.accessor$xRot((float) toRotation.y());
+                    packetInAccessor.accessor$yRot((float) toRotation.x());
+                }
             }
         }
     }
 
+    @Inject(
+            method = "handleMoveVehicle",
+            cancellable = true,
+            at = @At(value = "INVOKE", target = "Lnet/minecraft/server/level/ServerPlayer;getLevel()Lnet/minecraft/server/level/ServerLevel;")
+    )
+    private void impl$handleVehicleMoveEvent(final ServerboundMoveVehiclePacket param0, final CallbackInfo ci) {
+        final ServerboundMoveVehiclePacketAccessor packet = (ServerboundMoveVehiclePacketAccessor) param0;
+        final ServerPlayer player = (ServerPlayer) this.player;
+        final Entity rootVehicle = this.player.getRootVehicle();
+        final Vector3d fromRotation = new Vector3d(rootVehicle.yRot, rootVehicle.xRot, 0);
+
+        // Use the position of the last movement with an event or the current player position if never called
+        // We need this because we ignore very small position changes as to not spam as many move events.
+        final Vector3d fromPosition = VecHelper.toVector3d(rootVehicle.position());
+
+        final Vector3d originalToPosition = new Vector3d(param0.getX(), param0.getY(), param0.getZ());
+        final Vector3d originalToRotation = new Vector3d(param0.getYRot(), param0.getXRot(), 0);
+
+        // common checks and throws are done here.
+        final Tuple<Vector3d, Vector3d> events = this.impl$createAndFireMoveAndRotationEvents(
+                player,
```

Comment:
This should be rootVehicle, the player isn't moving, its only the passenger.

Reply:
Also pushing the player as cause here could be helpful so plugins can determine if player is controlling vehicle and link the player with it.

Reply:
You're right, this should be rootVehicle, good catch. Also, I think the player is on the stack already because they sent the packet.